python_scripts/metrics_sol_02.py

```diff
@@ -91,3 +91,43 @@
 }
 scores = pd.DataFrame(scores)
 scores
+
+# %% [markdown] tags=["solution"]
+# In the Regression Metrics notebook, we introduced the concept of loss function,
+# which is the metric optimized when training a model. In the case of a
+# `LinearRegression`, the fitting process consists in minimizing the mean squared
+# error (MSE). Some estimators, such as the `HistGradientBoostingRegressor`, can
+# use different loss functions, to be set using the `loss` hyperparameter.
+#
+# Notice that the evaluation metrics and the loss functions are not necessarily
+# the same. Let's see an example:
+
+# %%
+# solution
+from collections import defaultdict
+from sklearn.ensemble import HistGradientBoostingRegressor
+
+scoring = ["neg_mean_squared_error", "neg_mean_absolute_error"]
+loss_functions = ["squared_error", "absolute_error"]
+scores = defaultdict(list)
+
+for loss_func in loss_functions:
+    model = HistGradientBoostingRegressor(loss=loss_func)
+    cv_results = cross_validate(model, data, target, scoring=scoring)
+    mse = -cv_results["test_neg_mean_squared_error"]
+    mae = -cv_results["test_neg_mean_absolute_error"]
+    scores["loss"].append(loss_func)
+    scores["MSE"].append(f"{mse.mean():.1f} ± {mse.std():.1f}")
+    scores["MAE"].append(f"{mae.mean():.1f} ± {mae.std():.1f}")
+scores = pd.DataFrame(scores)
+scores.set_index("loss")
+
+# %% [markdown] tags=["solution"]
+# Even if the score distributions overlap due to the presence of outliers in the
+# dataset, it is true that the average MSE is lower when `loss="squared_error`,
+# whereas the average MAE is lower when `loss="absolute_error` as expected.
+# Indeed, the choice of a loss function is made depending on the evaluation
+# metric that we want to optimize for a given use case.
+#
```

Comment:
To address Guillaume remark above, we could probably insert a paragraph here along the lines:

> Note that it is not always possible to optimize directly some metrics.
> For instance, classification models are often evaluated using metrics
> computed on hard class predictions (e.g. accuracy or F1) rather
> than from the continuous values returned by the `predict_proba`
> or `decision_function` methods. Such hard class prediction metrics
> typically evolve in a discontinuous manner when progressively changing > the internal parameters of the model during training. Therefore they
> cannot be directly optimized by learning solvers based on gradient
> descent and the likes (the details are belong the scope of this MOOC).
> So instead classifiers are typically trained by optimizing a
> loss function computed from some continuous output of the model.
> For instance `LogisticRegression` minimizes the `log_loss` applied to
> the `predict_proba` output of the model. The `log_loss` acts
> as a "surrogate loss" for discontinuous classification metrics such as
> the accuracy: by minimizing the surrogate loss, we maximize their
> accuracy. However scikit-learn does not provide surrogate losses for
> all possible classification metrics.